packages/hash/api/src/lib/schemas/jsonSchema.ts

```diff
@@ -1,30 +1,62 @@
 import { JSONObject } from "blockprotocol";
-import Ajv2019 from "ajv/dist/2019";
+import Ajv2019, { AnySchemaObject } from "ajv/dist/2019";
 import addFormats from "ajv-formats";
+import { DataSource } from "apollo-datasource";
+import $RefParser, {
+  FileInfo,
+  JSONSchema,
+} from "@apidevtools/json-schema-ref-parser";
+import { cloneDeep } from "lodash";
 import { FRONTEND_URL } from "../config";
 
 /**
- * When compiling a schema AJV wants to resolve $refs to other schemas.
- * For now we can just give it empty schemas as the resolution for those $refs.
- * @todo check that $refs point to URIs which return at least valid JSON.
- *    We might not want to check each is a valid schema as they might link on to many more.
- *    For schemas stored in HASH, we know they're valid (since each is checked on insert).
+ * Given a list of all properties, check if any duplicates are present by the property name.
+ * @param allProps list of properties to duplicate check
+ * @returns a list of type-mismatch errors. Does not throw an exception.
  */
-const checkExternalSchemaExists = async (_uri: string) => {
-  return {};
-};
+const propertyKeyValidator = <T extends [x: string, _: any][]>(
+  ...allProps: T
+) => {
+  const seen: Map<string, any> = new Map([]);
+  const duplicateErrors: string[] = [];
 
-export const ajv = new Ajv2019({
-  addUsedSchema: false, // stop AJV trying to add compiled schemas to the instance
-  loadSchema: checkExternalSchemaExists,
-});
+  // Iterate all props, check for duplicates and throw an error if types mismatch
+  for (const [prop, value] of allProps) {
+    if (seen.has(prop)) {
+      const other = seen.get(prop);
 
-ajv.addKeyword({
-  keyword: "componentId",
-  schemaType: "string",
-});
+      if (value?.type !== other?.type) {
+        duplicateErrors.push(
+          `Type mismatch on "${prop}". Got "${value?.type}" expected "${other?.type}"`,
+        );
+      }
```

Comment:
I guess the type of thing I'm describing would violate the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) because an instance complying with the inheriting schema would be potentially incompatible with the parent.

Reply:
I was not aware of `"type": ["integer", "string"]` being possible in JSON Schema. I thought that was only handled through `oneOf`. 

I see how the relation 1 and 2 outlined here are compatible, I think we are able to make some ruleset of compatibility more advanced than just comparison using some subtyping hierarchy.  I just think that we should leave that for another task, as it might uncover more complex situations which would make this PR way bigger :)